src/aof.c

```diff
@@ -1184,99 +1615,174 @@ void backgroundRewriteDoneHandler(int exitcode, int bysignal) {
-
-        redisLog(REDIS_NOTICE,
-            "Parent diff successfully flushed to the rewritten AOF (%lu bytes)", aofRewriteBufferSize());
-
-        /* The only remaining thing to do is to rename the temporary file to
-         * the configured file and switch the file descriptor used to do AOF
-         * writes. We don't want close(2) or rename(2) calls to block the
-         * server on old file deletion.
-         *
-         * There are two possible scenarios:
-         *
-         * 1) AOF is DISABLED and this was a one time rewrite. The temporary
-         * file will be renamed to the configured file. When this file already
-         * exists, it will be unlinked, which may block the server.
-         *
-         * 2) AOF is ENABLED and the rewritten AOF will immediately start
-         * receiving writes. After the temporary file is renamed to the
-         * configured file, the original AOF file descriptor will be closed.
-         * Since this will be the last reference to that file, closing it
-         * causes the underlying file to be unlinked, which may block the
-         * server.
-         *
-         * To mitigate the blocking effect of the unlink operation (either
-         * caused by rename(2) in scenario 1, or by close(2) in scenario 2), we
-         * use a background thread to take care of this. First, we
-         * make scenario 1 identical to scenario 2 by opening the target file
-         * when it exists. The unlink operation after the rename(2) will then
-         * be executed upon calling close(2) for its descriptor. Everything to
-         * guarantee atomicity for this switch has already happened by then, so
-         * we don't care what the outcome or duration of that close operation
-         * is, as long as the file descriptor is released again. */
-        if (server.aof_fd == -1) {
-            /* AOF disabled */
+#ifdef USE_NVML
+        if (server.aof_use_nvml == 1) {
+            /*size_t real_log_rewr_size; */
+            if (server.aof_nvml_direct == 1) {
+                pmemlog_walk(server.aof_plp, 0, copyDiffPMEMlog, NULL);
+            } else {
+                if (aofRewriteBufferWritePMEMlog(server.aof_plp_rewr) == -1) {
+                    redisLog(REDIS_WARNING,
+                        "Error trying to flush the parent diff to the rewritten AOF: %s", strerror(errno));
+                    goto cleanup;
+                }
+            }
+            latencyEndMonitor(latency);
+            latencyAddSampleIfNeeded("aof-rewrite-diff-write",latency);
+
+            redisLog(REDIS_NOTICE,
+                "Parent diff successfully flushed to the rewritten AOF (%lu bytes)", aofRewriteBufferSize());
+
+            /* swap filenames, swap fd's  and plp's */
+            latencyStartMonitor(latency);
+            snprintf(tmpfile, 256, "aof-rewrite-swap-%d.aof", getpid());
+            if (rename (server.aof_filename, tmpfile) == -1) {
```

Comment:
Is this fail-safe?
What if the app crashes (or is killed) after the first rename()?